src/core/ext/filters/client_channel/resolver/google_c2p/google_c2p_resolver.cc

```diff
@@ -0,0 +1,360 @@
+                const grpc_http_response* response, grpc_error* error) override;
+  };
+
+  // A metadata server query to get the IPv6 address.
+  class IPv6Query : public MetadataQuery {
+   public:
+    IPv6Query(RefCountedPtr<GoogleCloud2ProdResolver> resolver,
+              grpc_polling_entity* pollent);
+
+   private:
+    void OnDone(GoogleCloud2ProdResolver* resolver,
+                const grpc_http_response* response, grpc_error* error) override;
+  };
+
+  void ZoneQueryDone(std::string zone);
+  void IPv6QueryDone(bool ipv6_supported);
+  void StartXdsResolver();
+
+  std::shared_ptr<WorkSerializer> work_serializer_;
+  grpc_polling_entity pollent_;
+  bool using_dns_ = false;
+  OrphanablePtr<Resolver> child_resolver_;
+
+  OrphanablePtr<ZoneQuery> zone_query_;
+  absl::optional<std::string> zone_;
+
+  OrphanablePtr<IPv6Query> ipv6_query_;
+  absl::optional<bool> supports_ipv6_;
+};
+
+//
+// GoogleCloud2ProdResolver::MetadataQuery
+//
+
+GoogleCloud2ProdResolver::MetadataQuery::MetadataQuery(
+    RefCountedPtr<GoogleCloud2ProdResolver> resolver, const char* path,
+    grpc_polling_entity* pollent)
+    : resolver_(std::move(resolver)) {
+  grpc_httpcli_context_init(&context_);
+  // Start HTTP request.
+  GRPC_CLOSURE_INIT(&on_done_, OnHttpRequestDone, this, nullptr);
+  Ref().release();  // Ref held by callback.
+  grpc_httpcli_request request;
+  memset(&request, 0, sizeof(grpc_httpcli_request));
+  grpc_http_header header = {const_cast<char*>("Metadata-Flavor"),
+                             const_cast<char*>("Google")};
+  request.host = const_cast<char*>("metadata.google.internal");
+  request.http.path = const_cast<char*>(path);
+  request.http.hdr_count = 1;
+  request.http.hdrs = &header;
+  grpc_resource_quota* resource_quota =
+      grpc_resource_quota_create("c2p_resolver");
+  grpc_httpcli_get(&context_, pollent, resource_quota, &request,
+                   ExecCtx::Get()->Now() + 10000,  // 10s timeout
+                   &on_done_, &response_);
+  grpc_resource_quota_unref_internal(resource_quota);
+}
+
+GoogleCloud2ProdResolver::MetadataQuery::~MetadataQuery() {
+  grpc_httpcli_context_destroy(&context_);
+}
+
+void GoogleCloud2ProdResolver::MetadataQuery::Orphan() {
+  // TODO(roth): Once the HTTP client library supports cancellation,
+  // use that here.
+  MaybeCallOnDone(GRPC_ERROR_CANCELLED);
+  Unref();
+}
+
+void GoogleCloud2ProdResolver::MetadataQuery::OnHttpRequestDone(
+    void* arg, grpc_error* error) {
+  auto* self = static_cast<MetadataQuery*>(arg);
+  self->MaybeCallOnDone(error);
+  grpc_http_response_destroy(&self->response_);
+  self->Unref();
+}
+
+void GoogleCloud2ProdResolver::MetadataQuery::MaybeCallOnDone(
+    grpc_error* error) {
+  bool expected = false;
+  if (on_done_called_.CompareExchangeStrong(
+          &expected, true, MemoryOrder::RELAXED, MemoryOrder::RELAXED)) {
+    // Hop back into WorkSerializer.
+    Ref().release();  // Ref held by callback.
```

Comment:
If I'm correct, we need to ref `error` here, since it's only current ref is the `OnHttpRequestDone` closure, which may finish before the work serializer closure runs?